File: LibraryManagementSystem_Java/LibraryManagementSystem/src/main/java/assignment/librarymanagementsystem/Item.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package assignment.librarymanagementsystem;

/**
 *
 * @author moonm
 */
abstract class Item 
{
    private String id;
    private int pageCount;
    private int timesBorrowed;
    private int totalCopies;
    private boolean isDamaged;
    private boolean isBorrowed;
    private DigitalCopy digitalCopy;

    public Item(String id, int pageCount, int totalCopies) 
    {
        this.id = id;
        this.pageCount = pageCount;
        this.totalCopies = totalCopies;
        this.timesBorrowed = 0;
        this.isDamaged = false;
        this.isBorrowed = false;
        this.digitalCopy = null;
    }

    public String getId() { return id; }
    public int getPageCount() { return pageCount; }
    public int getTimesBorrowed() { return timesBorrowed; }
    public int getTotalCopies() { return totalCopies; }
    public boolean getIsDamaged() { return isDamaged; }
    public boolean getIsBorrowed() { return isBorrowed; }
    public DigitalCopy getDigitalCopy() { return digitalCopy; }

    public void setPageCount(int pageCount) { this.pageCount = pageCount; }
    public void setTimesBorrowed(int timesBorrowed) { this.timesBorrowed = timesBorrowed; }
    public void setTotalCopies(int totalCopies) { this.totalCopies = totalCopies; }
    public void setIsDamaged(boolean isDamaged) { this.isDamaged = isDamaged; }
    public void setIsBorrowed(boolean isBorrowed) { this.isBorrowed = isBorrowed; }
    public void setDigitalCopy(DigitalCopy digitalCopy) { this.digitalCopy = digitalCopy; }

    /**
     * Checks if the item is damaged.
     * @return true if the item is damaged; false otherwise
     */
    public boolean ifDamaged() 
    {
        return isDamaged;
    }

    /**
     * Requests a repair for the item.
     * @return true if the item is damaged and needs repair
     */
    public boolean requestRepair() 
    {
        return isDamaged;
    }

    /**
     * Checks if the item is borrowed.
     * @return true if currently borrowed; false otherwise
     */
    public boolean ifBorrowed() 
    {
        return isBorrowed;
    }

    /**
     * Attempts to borrow the item if copies are available.
     * Updates borrowing state accordingly.
     * @return true if borrowing is successful; false if no copies are available
     */
    public boolean borrow() 
    {
        if (totalCopies > 0) 
        {
            totalCopies--;
            timesBorrowed++;
            if (totalCopies == 0) 
            {
                isBorrowed = true;
            }
            return true;
        }
        return false;
    }

    /**
     * Processes the return of an item and updates the damage and borrowing status.
     * @param damaged whether the item was returned damaged
     */
    public void returnItem(boolean damaged) 
    {
        totalCopies++;
        if (damaged) 
        {
            isDamaged = true;
        }
        if (totalCopies > 0) 
        {
            isBorrowed = false;
        }
    }

    /**
     * Creates a digital copy of this item if one doesn't already exist.
     */
    public void createDigitalCopy() 
    {
        if (digitalCopy == null) 
        {
            digitalCopy = new DigitalCopy(this.id);
        }
    }
}


class DigitalCopy 
{
    private String id;
    private int totalViews;
    private int totalDownloads;

    /**
     * Constructs a digital copy with a specific ID.
     * @param id Identifier of the original item
     */
    public DigitalCopy(String id) 
    {
        this.id = id;
        this.totalViews = 0;
        this.totalDownloads = 0;
    }

    /**
     * Increments the view counter for the digital copy.
     */
    public void view() 
    {
        totalViews++;
    }

    /**
     * Increments the download counter for the digital copy.
     */
    public void download() 
    {
        totalDownloads++;
    }

    public String getId() { return id; }
    public int getTotalViews() { return totalViews; }
    public int getTotalDownloads() { return totalDownloads; }
}

File: LibraryManagementSystem-Python/models/item.py
from abc import ABC, abstractmethod
from dataclasses import dataclass

@dataclass
class DigitalCopy:
    """
    Represents a digital copy of a library item.

    Attributes:
        id (str): Unique identifier for the digital copy
        total_views (int): Total number of views
        total_downloads (int): Total number of downloads
    """
    id: str
    total_views: int = 0
    total_downloads: int = 0

    def view(self):
        """Increment the total views of the digital copy."""
        self.total_views += 1

    def download(self):
        """Increment the total downloads of the digital copy."""
        self.total_downloads += 1

class Item(ABC):
    """
    Abstract base class representing a library item.
    
    Attributes:
        id (str): Unique identifier for the item
        page_count (int): Number of pages in the item
        total_copies (int): Total number of copies available
        times_borrowed (int): Number of times the item has been borrowed
        damage_severity (int): Level of damage from 0-10, 0 being undamaged
        is_damaged (bool): Whether the item is damaged
        is_borrowed (bool): Whether the item is currently borrowed
        digital_copy (DigitalCopy): Digital version of the item if available
    """

    def __init__(self, id: str, page_count: int, total_copies: int):
        """
        Initialize a new library item.

        Args:
            id (str): Unique identifier for the item
            page_count (int): Number of pages in the item
            total_copies (int): Total number of copies available
        """
        self.id = id
        self.page_count = page_count
        self.total_copies = total_copies
        self.times_borrowed = 0
        self.is_damaged = False
        self.is_borrowed = False
        self.digital_copy = None
        self.damage_severity = 0  # New attribute for damage severity

    def borrow(self) -> bool:
        """
        Borrow the item if it is available and not damaged.

        Returns:
            bool: True if the item was successfully borrowed

        Raises:
            ItemException: If the item is damaged or no copies are available
        """
        if self.is_damaged:
            raise ItemException("Cannot borrow damaged item")
        if self.total_copies <= 0:
            raise ItemException("No copies available")
        self.total_copies -= 1
        self.times_borrowed += 1
        if self.total_copies == 0:
            self.is_borrowed = True
        return True

    def return_item(self, damaged: bool, severity: int = 0):
        """
        Return the item to the library.

        Args:
            damaged (bool): Whether the item is returned damaged
            severity (int): Damage severity level if the item is damaged
        """
        self.total_copies += 1
        if damaged:
            self.set_damage_severity(severity)
        if self.total_copies > 0:
            self.is_borrowed = False

    def set_damage_severity(self, severity: int):
        """
        Set the damage severity level of the item.

        Args:
            severity (int): Damage level from 0-10

        Raises:
            ItemException: If severity is not between 0 and 10
        """
        if not isinstance(severity, int) or severity < 0 or severity > 10:
            raise ItemException("Damage severity must be between 0 and 10")
        self.damage_severity = severity
        self.is_damaged = severity > 0

    def get_damage_severity(self) -> int:
        """
        Get the damage severity level of the item.

        Returns:
            int: Damage severity level from 0-10
        """
        return self.damage_severity

    def create_digital_copy(self):
        """
        Create a digital copy of the item if it is not a Thesis.

        Raises:
            ItemException: If the item is a Thesis
        """
        if isinstance(self, Thesis):
            raise ItemException("Thesis items cannot have digital copies")
        if self.digital_copy is None:
            self.digital_copy = DigitalCopy(self.id)
